parlai/utils/conversations.py

```diff
@@ -0,0 +1,188 @@
+#!/usr/bin/env python3
+# Copyright (c) Facebook, Inc. and its affiliates.
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+
+"""
+Utility methods for conversations format.
+"""
+import datetime
+import json
+import os
+import random
+
+
+class Metadata:
+    """
+    Utility class for conversation metadata.
+
+    Metadata should be saved at <datapath>.metadata.
+    """
+
+    def __init__(self, datapath):
+        self._load(datapath)
+
+    def _load(self, datapath):
+        metadata_path = self._get_path(datapath)
+        if not os.path.isfile(metadata_path):
+            raise RuntimeError(
+                f'Metadata at path {metadata_path} not found. '
+                'Double check your path.'
+            )
+
+        with open(metadata_path, 'rb') as f:
+            metadata = json.load(f)
+
+        self.datetime = metadata['date']
+        self.opt = metadata['opt']
+        self.self_chat = metadata['self_chat']
+        self.speaker_1 = metadata['speaker_1']
+        self.speaker_2 = metadata['speaker_2']
+        self.extra_data = {}
+        for k, v in metadata.items():
+            if k not in ['date', 'opt', 'speaker_1', 'speaker_2']:
+                self.extra_data[k] = v
+
+    def read(self):
+        """
+        Read the relevant metadata.
+        """
+        print(f'Saved at: {self.datetime}')
+        print(f'Self chat: {self.self_chat}')
+        print(f'Speaker 1: {self.speaker_1}')
+        print(f'Speaker 2: {self.speaker_2}')
+        print('Opt:')
+        for k, v in self.opt.items():
+            print(f'\t{k}: {v}')
+        for k, v in self.extra_data.items():
+            print(f'{k}: {v}')
+
+    @staticmethod
+    def _get_path(datapath):
+        fle, _ = os.path.splitext(datapath)
+        return fle + '.metadata'
+
+    @classmethod
+    def save_metadata(
+        cls,
+        datapath,
+        opt,
+        self_chat=False,
+        speaker_1='human',
+        speaker_2=None,
+        **kwargs,
+    ):
+        """
+        Dump conversation metadata to file.
+        """
+        metadata = {}
+        metadata['date'] = str(datetime.datetime.now())
+        metadata['opt'] = opt
+        metadata['self_chat'] = self_chat
+        metadata['speaker_1'] = speaker_1
+        if speaker_2 is not None:
+            metadata['speaker_2'] = speaker_2
+        else:
+            metadata['speaker_2'] = opt.get('model')
+
+        for k, v in kwargs.items():
+            metadata[k] = metadata[v]
+
+        metadata_path = cls._get_path(datapath)
+        print(f'[ Writing metadata to file {metadata_path} ]')
+        with open(metadata_path, 'w') as f:
+            f.write(json.dumps(metadata))
+
+
+class Conversations:
```

Comment:
I'm wondering if the base level shouldn't be a `Conversation`, and conversations would fundamentally be a collection of objects (therefore implmenting `__len__`, `__iter__`, ...). Metadata would (as you have here) also be a part of Conversation.

The future I'm imagining is that the teacher code is also Conversation oriented. So when I need to create a new dataset from some arbitrary data, I need to parse out a single Conversation and emit it. Higher level logic could then turn a Conversation into a teacher, building the labels, handling the symmetry or not, even building label candidates automatically...

Building a conversation should be similarly easy, and grouping waves of them into a Conversation.

Probably we want to think this through more.

Reply:
so having another class called `Conversation` to replace the idea of episodes?

Reply:
The name `Conversation` makes more intuitive sense than `episode` anyway so @stephenroller might be onto something.

I kind of agree that we might want to go the `Conversation` route and have `Conversations` be a collection of `Conversation` objects. At the very least this would perhaps ease the logic around what conversation-level information is available, e.g. including the `personas` or other context (rather than just reading through the arbitrary keys in the dialogue).

Idk i do think this is good for now but could definitely warrant a more in-depth discussion